src/core/text/TextMetrics.js

```diff
@@ -0,0 +1,315 @@
+        const fontProperties = TextMetrics.measureFont(font);
+        const context = canvas.getContext('2d');
+
+        context.font = font;
+
+        const outputText = wordWrap ? TextMetrics.wordWrap(text, style, canvas) : text;
+        const lines = outputText.split(/(?:\r\n|\r|\n)/);
+        const lineWidths = new Array(lines.length);
+        let maxLineWidth = 0;
+
+        for (let i = 0; i < lines.length; i++)
+        {
+            const lineWidth = context.measureText(lines[i]).width + ((lines[i].length - 1) * style.letterSpacing);
+
+            lineWidths[i] = lineWidth;
+            maxLineWidth = Math.max(maxLineWidth, lineWidth);
+        }
+        let width = maxLineWidth + style.strokeThickness;
+
+        if (style.dropShadow)
+        {
+            width += style.dropShadowDistance;
+        }
+
+        const lineHeight = style.lineHeight || fontProperties.fontSize + style.strokeThickness;
+        let height = Math.max(lineHeight, fontProperties.fontSize + style.strokeThickness)
+            + ((lines.length - 1) * lineHeight);
+
+        if (style.dropShadow)
+        {
+            height += style.dropShadowDistance;
+        }
+
+        return new TextMetrics(
+            text,
+            style,
+            width,
+            height,
+            lines,
+            lineWidths,
+            lineHeight,
+            maxLineWidth,
+            fontProperties
+        );
+    }
+
+    /**
+     * Applies newlines to a string to have it optimally fit into the horizontal
+     * bounds set by the Text object's wordWrapWidth property.
+     *
+     * @private
+     * @param {string} text - String to apply word wrapping to
+     * @param {PIXI.TextStyle} style - the style to use when wrapping
+     * @param {HTMLCanvasElement} [canvas] - optional specification of the canvas to use for measuring.
+     * @return {string} New string with new lines applied where required
+     */
+    static wordWrap(text, style, canvas = TextMetrics._canvas)
+    {
+        const context = canvas.getContext('2d');
+
+        // Greedy wrapping algorithm that will wrap words as the line grows longer
+        // than its horizontal bounds.
+        let result = '';
+        const lines = text.split('\n');
+        const wordWrapWidth = style.wordWrapWidth;
+
+        for (let i = 0; i < lines.length; i++)
+        {
+            let spaceLeft = wordWrapWidth;
+            const words = lines[i].split(' ');
+
+            for (let j = 0; j < words.length; j++)
+            {
+                const wordWidth = context.measureText(words[j]).width;
+
+                if (style.breakWords && wordWidth > wordWrapWidth)
+                {
+                    // Word should be split in the middle
+                    const characters = words[j].split('');
+
+                    for (let c = 0; c < characters.length; c++)
+                    {
+                        const characterWidth = context.measureText(characters[c]).width;
```

Comment:
It seems like caching these characters makes some difference (9ms vs 4ms in my demo), especially if there's more duplicate characters..